Support multiple uid/gid mappings [1/2]

henry118 opened the pull request:
Enhance user namespace implementation to support multi-entry uid/gid mappings.

This is 1st patch (out of 2) that supports multiple entries of uid/gid configurations. This patch focuses on the "slow" path that requires `chown` on every file in the root FS. uidmapped mounts implementation will be covered in a separate PR.

This implementation has changes in the following area:

1. Ported the `IdentityMapping` implementation from `idtools` package in Moby.
2. Added a few container opt funcs to take multiple uid/gid mappings.
3. Allowing multiple `--uidmap, --gidmap` options in `ctr run` command.
4. Fixed a bug where special file permissions like setuid, setgid or sticky bits are lost during `chown`

An sample command to run a container with remapped `foo` user:

```
ctr run --uidmap 0:100000:1000 --uidmap 1000:201795:1 --uidmap 1001:10000000:64535 --gidmap 0:100000:100 --gidmap 100:1001:1 --gidmap 101:10000000:65435 --user foo <image> test bash
```

I hope the implementation makes sense and I'm looking forward to your feedback :)

Kern-- commented on client/snapshotter_opts_unix.go:
```diff
@@ -110,7 +111,11 @@ func resolveSnapshotOptions(ctx context.Context, client *Client, snapshotterName
 		return "", err
 	}
 	// TODO(dgl): length isn't taken into account here yet either.
```
Since you're already touching this, does it make sense to update the snapshot naming here and in `container_opts_unix.go` to encode the full idmap rather than just the root user map? 

I'm not sure how likely it would be, but you could get the wrong result if two configs on the same host have the same root mapping, but different non-root mappings.

henry118 replied:
The updated version will use the hash of (parent, idmap) for the snapshot name. That way should address this issue.

estesp commented:
/test pull-containerd-node-e2e

estesp commented:
I'm adding to 2.0 milestone; @containerd/maintainers let me know if anyone thinks it should not be in 2.0

estesp commented on cmd/ctr/commands/run/run_unix.go:
```diff
@@ -45,11 +47,11 @@ import (
-	&cli.StringFlag{
+	&cli.StringSliceFlag{
 		Name:  "uidmap",
 		Usage: "Run inside a user namespace with the specified UID mapping range; specified with the format `container-uid:host-uid:length`",
```
I think the usage string (for this and gidmap) needs to reflect that it is now a slice/list of maps; at least the first clause should end with **mapping range(s)**; could also add at the end a new clause: ", separate multiple mappings with a comma"

henry118 replied:
After this change, the help message for this option becomes the following. 
```
--uidmap container-uid:host-uid:length [ --uidmap container-uid:host-uid:length ]  Run inside a user namespace with the specified UID mapping range; specified with the format container-uid:host-uid:length
```
I think the additional `[]` will indicate the option is a slice. But I think updating the "mapping range" to "mapping ranges" was a good call out. I will make the update in the next revision.

henry118 commented:
/test pull-containerd-node-e2e

AkihiroSuda commented:
@henry118 Do you plan to add the second part in this PR? Or will it be a separate PR?

estesp commented:
> @henry118 Do you plan to add the second part in this PR? Or will it be a separate PR?

He was going to do it as a separate PR; he is out right now so it will be after he returns

henry118 commented:
@AkihiroSuda I'll add the 2nd part of the PR after this one, as I'd like keep the patch small. I also updated the PR to include the test. PTAL. Thanks!

AkihiroSuda commented:
cc @rata PTAL

rata reviewed (commented):
I had a quick look, left some comments.

I guess this would work with CRI if it specifies more than one mapping? Or is there some glue missing?

henry118 commented on client/container_opts_unix.go:
```diff
@@ -95,22 +118,30 @@ func withRemappedSnapshotBase(id string, i Image, uid, gid uint32, readonly bool
+		if cerr = os.Lchown(path, int(h.UID), int(h.GID)); cerr != nil {
+			return cerr
+		}
+		// we must retain special permissions such as setuid, setgid and sticky bits
+		if mode := info.Mode(); mode&os.ModeSymlink == 0 && mode&(os.ModeSetuid|os.ModeSetgid|os.ModeSticky) != 0 {
+			return os.Chmod(path, mode)
```
I don't think we have test images for this use case yet. The way I tested it was configuring a base ubuntu image with sudoers and tested the sudo command. In fact this also fixed a bug in the current implementation in the sudo scenario. Can I postpone this effort to a follow-up PR? We may need to discuss if adding an another test image is ok for this use case.

rata replied:
@henry118 are you sure we need a new image? I think the sticky bit should be present on /tmp on most container images. It seems busybox has it, trying locally.

Or maybe a unit test is also helpful.

henry118 replied:
@rata Sorry while I was talking about sticky bit, I was thinking about setuid/setgid bits on files like `/usr/bin/passwd` or `/usr/bin/sudo`, etc. On alpine/busybox, those files do not exist or do not have the bits set. But for sticky bits, it was successfully retained by `chown` even without my patch based on testing. So testing with a different image (i.e., ubuntu) is the only way that I can think of. 

As to UT, none of the units under `client` have a corresponding unit test today. I'm not sure if I should add one just for this test case.

Happy to hear other suggestions.

henry118 replied:
Oh, it looks like the `ghcr.io/containerd/volume-ownership:2.1` image we use for volume ownership tests may have those files we need. I'll see if I can use that image to test the setuid/git bits.

henry118 replied:
OK, I updated the integ test to cover this case. PTAL

rata replied:
Thanks!

henry118 commented on client/snapshotter_opts_unix.go:
```diff
@@ -120,3 +133,27 @@ func resolveSnapshotOptions(ctx context.Context, client *Client, snapshotterName
+
+type remappedSnapshot struct {
+	Parent string              `json:"Parent"`
+	IDMaps sys.IdentityMapping `json:"IDMaps"`
```
Yeah this is a bit overloaded name. To make it clearer, I moved my implementation to a new userns package under sys, and renamed the struct name to IDMap, which is I believe is more concise and also aligned with the moby code [here](https://github.com/moby/sys/blob/main/user/user.go#L53-L57). Hopefully it makes the code cleaner.

henry118 commented on cmd/ctr/commands/run/run_unix.go:
```diff
@@ -159,26 +161,47 @@ func NewContainer(ctx context.Context, client *containerd.Client, cliContext *cl
+					if err != nil {
+						return nil, err
+					}
+					uidSpec = append(uidSpec, uidMap)
```
I refactored the logic into a func. PTAL

henry118 commented on integration/client/container_linux_test.go:
```diff
@@ -1222,6 +1224,113 @@ func testUserNamespaces(t *testing.T, readonlyRootFS bool) {
 	}
 }
 
+func testMultiIDMappedUserNamespaces(t *testing.T, readonlyRootFS bool) {
```
Updated

rata replied:
Did you forget to push this? I don't see any changes to that file

rata replied:
Okay, the "updated" comment seem like you updated to use that. Sorry I didn't get it.

henry118 commented on pkg/sys/userns_idmap.go:
```diff
@@ -0,0 +1,124 @@
+/*
+   Copyright The containerd Authors.
```
Added comment to reference the original file

rata replied:
I don't know if that is valid. I'd like other maintainers input on this

estesp replied:
Yes, this is a fairly common practice. We've had copies of Go internal code (tar functionality?), Docker has as well, I believe Kubernetes has copied files, too. You can't just change the license, but you can add your own copyright header since it is part of our codebase now. I wrote the original package in Moby anyway and I am a containerd author, so there's that too :)

rata replied:
Thanks!

henry118 commented:
> I had a quick look, left some comments.
> 
> I guess this would work with CRI if it specifies more than one mapping? Or is there some glue missing?

Thanks for reviewing. I updated the PR to address most of your comments. PTAL.

In terms of CRI, it currently executes [the ID mapped mounts code path](https://github.com/containerd/containerd/blob/main/internal/cri/server/helpers_linux.go#L206), which is not updated by this PR. But it will be covered by my 2nd part of the patch.

rata commented on cmd/ctr/commands/run/run_unix.go:
```diff
@@ -159,26 +161,47 @@ func NewContainer(ctx context.Context, client *containerd.Client, cliContext *cl
 				if cliContext.Bool("remap-labels") {
-					cOpts = append(cOpts, containerd.WithNewSnapshot(id, image,
-						containerd.WithRemapperLabels(0, uidMap.HostID, 0, gidMap.HostID, uidMap.Size)))
+					// TODO: the optimization code path on id mapped mounts only supports single
+					// mapping entry today. use the root pair in this scenario for now.
```
What do you mean with the existing behavior? Today if you specify more than one mapping, however you manage to do that, an error is returned, right?

With this, we ignore the rest of the mappings, just take the mapping the root user in the container.

Am I missing something? Why can't we return an error if we can0t honor the request?

henry118 replied:
Sure, updated

rata replied:
Thanks!

henry118 commented:
/test pull-containerd-node-e2e

henry118 commented:
@rata can you take another look? thanks!

rata commented:
@henry118 
> In terms of CRI, it currently executes [the ID mapped mounts code path](https://github.com/containerd/containerd/blob/main/internal/cri/server/helpers_linux.go#L206), which is not updated by this PR. But it will be covered by my 2nd part of the patch.

CRI should only use idmap mounts if the kernel and the snapshotter support it. If you are using an old kernel, it will use chown. This is indeed what is happening in CI when the kernel is old.


If this is targeted for 2.0, I'm unsure we want to do this change now for CRI, though. What do other maintainers think?

rata commented on integration/client/container_linux_test.go:
```diff
@@ -1091,9 +1093,24 @@ func TestContainerKillInitPidHost(t *testing.T) {
+		uidmaps        = []specs.LinuxIDMapping{{ContainerID: 0, HostID: 1000, Size: 65535}}
+		gidmaps        = []specs.LinuxIDMapping{{ContainerID: 0, HostID: 2000, Size: 65535}}
+		cmdExit        = withExitStatus(7)
+		cmdCheckSetuid = withProcessArgs("bash", "-c", "[ -u /usr/bin/passwd ] && exit 7")
```
I'd add a comment to explain this 7. It's not obvious.

henry118 replied:
7 was the value used in the existing tests. I simply reused it. I guess it's just a random value to differentiate from common exit code like 0 or 1.

rata replied:
I know, but it is not so obvious now. I think a comment will help

henry118 replied:
OK, I refactored that to table tests, which should be self explanatory. Does it look better now?

rata replied:
Sorry the confusion, I was just saying to add a comment there. But yes, table test do look way nicer here, IMHO, so please keep them.

But can you add a comment on top of the lines that add the return 7, explaining that `testUserNamespaces()` uses that return code, etc.? Just a one-line comment should be enough, IMHO.

henry118 replied:
Ah okey. updated.

rata commented on pkg/sys/userns/idmap.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// RootID returns the ID pair for the root user
+func (i IDMap) RootPair() (User, error) {
```
Now this is unused, right?

rata replied:
When would this be used? For idmap mounts, we want to either honor the mappings or fail. This is to not honor the mappings and find the one for root. When will we use this?

henry118 replied:
Removed

rata replied:
Thanks!

rata commented on pkg/sys/userns/idmap.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// ToContainer returns the container Identify pair for the host uid and gid
+func (i IDMap) ToContainer(pair User) (User, error) {
```
This seems unused too

henry118 replied:
Same as https://github.com/containerd/containerd/pull/10307#discussion_r1731539158

rata replied:
Again, when will be used? If you have a follow-up PR coming that will use it, can't we add it there?

henry118 replied:
Removed

rata replied:
Thanks!

henry118 commented:
@rata 

> CRI should only use idmap mounts if the kernel and the snapshotter support it. If you are using an old kernel, it will use chown. This is indeed what is happening in CI when the kernel is old.

Yes I am aware of the fallback. The fallback is currently implemented in `resolveSnapshotOptions`. The entry point of this func is based on the idmapping labels configured on snapshotters, which is still limited to single mapping entry without my 2nd PR.

https://github.com/containerd/containerd/blob/7b948faa084506e72967d88b5d35423211f13b07/client/snapshotter_opts_unix.go#L113

> If this is targeted for 2.0, I'm unsure we want to do this change now for CRI, though. What do other maintainers think?

We can discuss about how we want to support CRI in my next PR that wires up the CRI logic. Basically if we don't want to change the CRI logic, we can simply keep using the old API in CRI. But I don't think the discussion is relevant to this PR though.

https://github.com/containerd/containerd/blob/7b948faa084506e72967d88b5d35423211f13b07/internal/cri/server/helpers_linux.go#L206

rata commented:
@henry118 okay, you were saying the opposite before (CRI userns required idmap mounts) :)

If you plan to do a follow-up PR with the idmap mounts part and prefer to wire CRI there, it seems fine to me. I think the final release should be close and changing that later seems safer, IMHO, besides other PRs that are inflight that would conflict with this. But I'm not a maintainer here :)

henry118 commented:
@rata 

> I think the final release should be close and changing that later seems safer, IMHO, besides other PRs that are inflight that would conflict with this. But I'm not a maintainer here :)

My 2nd PR is pretty much ready. I didn't open it because GH does not support stacked PRs. And I wanted to use the 1st PR to align on all the APIs. I wish that I can still catch the 2.0 release window to get this feature in. But I agree it's up to the maintainers. Again let's defer this discussion to the follow-up patch.

I believe I should have addressed all your comments at this stage. PTAL and let me know what else I can do to move this further.

rata commented:
@henry118 cool, thanks! Just to understand, would you like this PR for 2.0? Or also the follow-up PR?

I think maintainers need to review now, that is the next step :)

henry118 commented:
Ping @estesp @AkihiroSuda @dmcgowan PTAL. Really hope this enhancement can be part of 2.0.

estesp reviewed (approved):
LGTM

estesp commented:
@dmcgowan would love your review and yours or any @containerd/committers opinion on 2.0 inclusion. It improves our general user namespaces support to handle ranges, which is a core feature of user namespaces in Linux and has a follow-on change for the CRI as well that @henry118 has ready to go.

cpuguy83 commented on cmd/ctr/commands/run/run_unix.go:
```diff
@@ -159,26 +160,28 @@ func NewContainer(ctx context.Context, client *containerd.Client, cliContext *cl
-						containerd.WithRemapperLabels(0, uidMap.HostID, 0, gidMap.HostID, uidMap.Size)))
+					// TODO: the optimization code path on id mapped mounts only supports single mapping entry today.
+					if len(uidSpec) > 1 || len(gidSpec) > 1 {
+						return nil, errors.New("'remap-labels' option only supports a single mapping entry today")
```
Can you reword this to something to the effect of `'remap-labels' does not support multiple mappings`

henry118 replied:
Updated. PTAL

cpuguy83 reviewed (changes requested):
I think I'd prefer the idmap type to be an internal type.
Other than that this seems OK
